### DISCUSSION TITLE
tree-sitter: Failed to execute C compiler, No such file or directory

### DISCUSSION THREAD

**Discussion by asumbek** [tree-sitter/tree-sitter#2106] [Q&A]
Hello, It's my first time using tree-sitter. I tried creating grammar by following this [tree-sitter page](https://tree-sitter.github.io/tree-sitter/creating-parsers#project-setup). But failed in step parsing a file.

The error of it:
```
Failed to load language in current directory

Caused by:
    0: Failed to execute C compiler
    1: No such file or directory (os error 2)
```

GCC and Clang version:
```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib64/gcc/x86_64-suse-linux/12/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-suse-linux
Configured with: ../configure --prefix=/usr --infodir=/usr/share/info --mandir=/usr/share/man --libdir=/usr/lib64 --libexecdir=/usr/lib64 --enable-languages=c,c++,objc,fortran,obj-c++,ada,go,d,jit --enable-offload-targets=nvptx-none,amdgcn-amdhsa, --enable-offload-defaulted --without-cuda-driver --enable-host-shared --enable-checking=release --disable-werror --with-gxx-include-dir=/usr/include/c++/12 --enable-ssp --disable-libssp --disable-libvtv --enable-cet=auto --disable-libcc1 --enable-plugin --with-bugurl=https://bugs.opensuse.org/ --with-pkgversion='SUSE Linux' --with-slibdir=/lib64 --with-system-zlib --enable-libstdcxx-allocator=new --disable-libstdcxx-pch --enable-libphobos --enable-version-specific-runtime-libs --with-gcc-major-version-only --enable-linker-build-id --enable-linux-futex --enable-gnu-indirect-function --program-suffix=-12 --without-system-libunwind --enable-multilib --with-arch-32=x86-64 --with-tune=generic --with-build-config=bootstrap-lto-lean --enable-link-mutex --build=x86_64-suse-linux --host=x86_64-suse-linux
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.2.1 20230124 [revision 193f7e62815b4089dfaed4c2bd34fd4f10209e27] (SUSE Linux)

clang version 15.0.7
Target: x86_64-suse-linux
Thread model: posix
InstalledDir: /usr/bin
```

Directory list:
```
binding.gyp  bindings  Cargo.toml  example-file  grammar.js  LICENSE  node_modules  package.json  pnpm-lock.yaml  README.md  src
```

Step I do:
* `gh repo create`
* `cd tree-sitter-arturo`
* `pnpm i nan && pnpm i tree-sitter-cli -D`
* `nvim grammar.js`
* `pnpm exec tree-sitter generate`
* `echo hello > example-file`
* `pnpm exec tree-sitter parse example-file` This where the error begins

**Comment by ahlinc** [comment]
~I suspect you don't have an installed compiler...~, sorry, I overlooked the output.

Could you describe more your environment? Do you have a repo for a grammar that you tried to build?

Could you try to adapt to your grammar compiler calls that mentioned in a bellow message?

**Comment by sogaiu** [comment]
The original post seemed to suggest that gcc and clang are available -- may be I misread that -- or perhaps they are not available on `PATH`?

FWIW, when I had build issues before I modified code near: https://github.com/tree-sitter/tree-sitter/blob/c51896d32dcc11a38e41f36e3deb1a6a9c4f4b14/cli/loader/src/lib.rs#L427 to get a dump of the actual compilation command that is executed by the `tree-sitter` cli.  That allowed me to manually perform the invocation and experiment / diganose more easily.

May be it's worth having a debug flag to echo the invocation for investigative purposes?

**Reply by ahlinc** [reply]
@sogaiu in a temporary tree-sitter fork discussed in a Discord I've added a `TREE_SITTER_VERBOSE` env variable that outputs actual compilation commands.
It can be set like:
```sh
export TREE_SITTER_VERBOSE=1
```
And get an output by:
```sh
tree-sitter generate --build
```

**Comment by sogaiu** [comment]
@ahlinc Nice!

Gave it a try:
```
$ cd ~/src/tree-sitter-clojure
$ TREE_SITTER_VERBOSE=1 TREE_SITTER_DIR=$(pwd)/.tree-sitter tree-sitter.ahlinc generate --build --no-bindings --abi 13
cc -O2 -ffunction-sections -fdata-sections -fPIC -m64 -I /home/user/src/tree-sitter-clojure/src -Wall -Wextra -fno-exceptions -o /home/user/src/tree-sitter-clojure/src/parser.o -c /home/user/src/tree-sitter-clojure/src/parser.c '-DTREE_SITTER_INTERNAL_BUILD=1'
cc -O2 -ffunction-sections -fdata-sections -fPIC -m64 -shared -Wall -Wextra -fno-exceptions -o /home/user/.cache/tree-sitter/lib/clojure.so /home/user/src/tree-sitter-clojure/src/parser.o
```
I especially like the `-o` part so I can see the path of the resulting shared object / dynamic library.

Thanks!

**Reply by ahlinc** [reply]
As you can see in the output the `-o` parameters are present on every compiler call, it's needed to make a separate unit compilation per file (original tree-sitter compiles all files as a one unit), this is a first step to solve existing compiling issues when for ex. a c++ scanner code requires custom compiling flags or when scanner would be consist from several files, etc.

**Comment by sogaiu** [comment]
Yes, I get that `-o` is often present in compiler invocations -- as I understand it though, one does not always get a full path.  It depends on the how the invocation is constructed.

Was just happy to be able to see exactly where the shared library ended up because in the past I had encountered issues where the shared library being used was not the one I thought (because of the change in the location that gets referenced `~/.tree-sitter/bin/` vs `~/.cache/tree-sitter/lib`).

In any case, thanks again for this :)

---

_Update_: [this PR](https://github.com/tree-sitter/tree-sitter/pull/4645) made it possible to specify the exact shared library being used.  Hopefully the feature will be in a released version of the cli before too long.

**Comment by Nadelio** [comment]
Does tree-sitter only work with Clang? I can't get it to recognize my gcc compiler at all. I just get this error. (Name censored in file paths)
```
Failed to load langauge for path "example.niva"

Caused by:
    0: Failed to load language in current directory
    1: Failed to execute the C compiler with the following command:
       "clang" "-O2" "-ffunction-sections" "-fdata-sections" "-m64" "--target=x86_64-pc-windows-msvc" "-std=c11" "-I" "C:\\Users\\#####\\tree-sitter-niva\\.\\src" "-Wall" "-Werror=implicit-function-declaration" "-shared" "C:\\Users\\#####\\tree-sitter-niva\\.\\src\\parser.c" "-o" "C:\\Users\\#####\\AppData\\Local\\tree-sitter\\lib\\niva.dll"
    2: program not found
```

I am in the root working directory of my parser, I'm on the `Generate` step of the tree sitter website.

**Reply by ObserverOfTime** [reply]
GCC doesn't support the `-msvc` target. You have to set it to `-gnu`.
